R/estimateAlpha.R

```diff
@@ -0,0 +1,245 @@
+            index[i] <- gsub("_richness", "", index[i])
+            FUN <- .estimate_richness
+        } else {
+            stop("'index' is coresponding to none of the alpha diversity indices.",
+                 call. = FALSE)
+        }
+        # Performing rarefaction if rarefaction.depth is specified to be less 
+        # the max of the total read counts
+        if (rarefaction.depth < max(colSums(assay(x, assay.type)), na.rm = TRUE)) {
+            x <- .alpha_rarefaction(x, n.iter = n.iter, seed = seed,
+                                    args.sub = list(assay.type=assay.type,
+                                                    min_size=rarefaction.depth,
+                                                    verbose=FALSE),
+                                    FUN=FUN,
+                                    args.fun=list(index=index[i], assay.type="subsampled"),
+                                    ...,
+                                    name=name[i])
+        } else {
+            suppressWarnings(x <- do.call(FUN, args = c(list(x, assay.type=assay.type,
+                                                             index=index[i],
+                                                             name=name[i]),
+                                                        list(...))))
+        }
+    }
+    return(x)
+}
+
+## Helper functions
+
+.get_indices <- function(measure) {
+    switch(measure,
+           "diversity" = c("coverage_diversity", "faith_diversity", 
+                           "fisher_diversity", "gini_simpson_diversity",
+                           "inverse_simpson_diversity",
+                           "log_modulo_skewness_diversity", "shannon_diversity"),
+           "dominance" = c("absolute_dominance",
+                           "dbp_dominance", "core_abundance_dominance",
+                           "gini_dominance", "dmn_dominance", "relative_dominance",
+                           "simpson_lambda_dominance"),
+           "evenness" = c("camargo_evenness", "pielou_evenness", "simpson_evenness",
+                          "evar_evenness", "bulla_evenness"),
+           "richness" = c("ace_richness", "chao1_richness", "hill_richness", 
+                          "observed_richness"))
+}
+
+.alpha_rarefaction <- function(x,
+                               n.iter=1L,
+                               seed=123,
+                               args.sub=list(assay.type="counts",
+                                             min_size=min(colSums(assay(x, "counts")),
+                                                          na.rm = TRUE),
+                                             verbose=FALSE),
+                               FUN=.estimate_diversity,
+                               args.fun=c(index="shannon",
+                                          assay.type="subsampled"),
+                               ...,
+                               name = args.fun$index) {
+    set.seed(seed)
```

Comment:
seed still included despite the discussion otherwise?

Reply:
Once again, the earlier discussion resulted in keeping the seed for reproducibility? the subsampling function does it internally for each iteration, so keeping a seed outside the function call will enable the user to reproduce results....?

Reply:
Seed can be given outside the function and the results remain reproducible. 

Bioconductor recommends not having an additional seed in the functions.

I do not see why we need the seed inside function for reproducibility. You can test this yourself: apply the function two times without setting seed outside -> different results; set seed outside and immediately run the function -> same results every time

Reply:
ok I checked the subsampling function, I see now that it would produce the same seed over and over. But then results are not reproducible for the whole process due to [link](https://github.com/microbiome/mia/blob/master/R/subsampleCounts.R#L87C28-L87C61) ... but guess that's the problem of the subsampling function. I remove it from here.

Reply:
Yes, let us set seed outside of function if necessary, and remove from the function itself. Then the function result will remain reproducible.

This should always generate the same:

```
set.seed(24452)
estimateAlpha(...)
```

If this is not the case, prove me wrong.

Reply:
I already said setting outside the function is reproducible. "Outside the function" was not clear at comments for me....
but setting one seed before the lapply is not going to make the rarefaction happen since same subsampling results will be produced in every iteration.....

Reply:
That's why the communication was confusing after asking at the beginning to set one seed before the lapply...

Reply:
I do not think that statment is true: if you set seed outside the function, and then run lapply within the function, the lapply will not produce the same subsampling results on every iteration. If it does, that is unexpected and I would like to get a reproducible example that demonstrates that.

Reply:
Or perhaps I have cognitive problems too, which does not make me a good candidate for being a developer... IDK it's possible...

Reply:
You can try it out: run lapply() in those alternative ways:

# One general seed before lapply
1) 
set.seed(123); res1a <- lapply(....)
set.seed(123); res1b <- lapply(....)
-> res1a and res1b will yield identical outcomes -> result is reproducible

# Same seed initialized again at each iteration
2) res2 <- lapply(...., function (x) {set.seed(123); ... })
-> this will yield different outcome than res1a or res1b because here the seed is reset to 123 at each new iteration instead of just continuing the seedings from where the previous iteration left

# Random seed per each iteration
3) res3 <- lapply(...., function (x) {set.seed(sample(1e6, 1)); ... })
-> this yields different outcome from res1a, res1b, res2 since it uses different (randomly generated) seed at every iteration

Reply:
> I do not think that statment is true: if you set seed outside the function, and then run lapply within the function, the lapply will not produce the same subsampling results on every iteration. If it does, that is unexpected and I would like to get a reproducible example that demonstrates that.

Yes it won't, it will produced only the first iteration the same and the rest is not reproducible, and the question I had in mind is how would you make the whole iterations reproducible to have final reproducible results. That's why I still do think that generating a list of seeds and passing on each iteration the corresonding seed to subsampleCount is simple, safe and definitely reproducible...

Reply:
ok, I see your point now, I didn't know that setting one seed before lapply will make reproducible result:
![image](https://github.com/microbiome/mia/assets/86964562/ac2d90eb-f09a-454c-9b42-edc0c383147c)

Reply:
Yes, this is what I said: the result for the entire loop is reproducible when the seed it set before applying the loop.

Reply:
Is this now clear?

Reply:
Yes it is, I tested also outside the function and gives same results.
But what's honestly not clear is why you asked me to put it before the lapply at the begining then re-asked to remove it, and Tuomas also was ok with both solutions and you didn't comment? :) I have noticed this multiple times and it's the same story in every PR I make... I am not really motivated for this especially when I have other tasks at hand waiting for me to be done...  It's ok I understand I am not good enough of a developer, so perhaps I should do smt else instead? or go to coding 101. Maybe I could just open issues when encounter some and come here when I am good enough someday?

Reply:
The discussion evolved. 

Initially I thought that before lapply is enough. Then Tuomas pointed out that Bioc does not recommend including seed in the function. It sounded reasonable, so I changed my thinking and concluded that we can remove the seed from the function entirely. I might have not communicated that clearly in a hurry as I mistakenly thought I stated it in the later comments that the seed should be placed outside of the function.

The other points we should discuss elsewhere than in the public Github issues, thanks.

Reply:
My apologies, if I have not expressed the task with sufficient clarity, will try to improve on that now.